GVFS/GVFS/RepairJobs/GitRefsRepairJob.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            IEnumerable<string> badRefs = this.GetRefs().Where(x => !this.TryParseRef(x));
+
+            int numFailures = 0;
+
+            foreach (string badRef in badRefs)
+            {
+                if (!this.TryWriteRefFromLog(badRef, messages))
+                {
+                    numFailures++;
+                }
+            }
+
+            if (numFailures > 0)
+            {
+                messages.Add($"Not all references could be fixed. Failed to fix {numFailures} references.");
+                return FixResult.Failure;
+            }
+
+            return FixResult.Success;
+        }
+
+        /// <summary>
+        /// Get the list of full symbolic references (eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master)
+        /// to inspect for and correct issues.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract IEnumerable<string> GetRefs();
+
+        /// <summary>
+        /// Check if the contents of a reference is valid.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        /// <param name="refContents">Contents of the ref file</param>
+        protected virtual bool IsValidRefContents(string fullSymbolicRef, string refContents)
+        {
+            return SHA1Util.IsValidShaFormat(refContents);
+        }
+
+        /// <summary>
+        /// 'git ref-log' doesn't work if the repo is corrupted, so parsing reflogs seems like the only solution.
+        /// </summary>
+        /// <param name="fullSymbolicRef">A full symbolic ref name. eg. HEAD, refs/remotes/origin/HEAD, refs/heads/master</param>
+        private static bool TryReadLastRefLogEntry(Enlistment enlistment, string fullSymbolicRef, out RefLogEntry refLog, out string error)
+        {
+            string refLogPath = Path.Combine(enlistment.WorkingDirectoryRoot, GVFSConstants.DotGit.Logs.Root, fullSymbolicRef);
+            if (!File.Exists(refLogPath))
+            {
+                refLog = null;
+                error = $"Could not find reflog for ref '{fullSymbolicRef}'";
+                return false;
+            }
+
+            try
+            {
+                string refLogContents = File.ReadLines(refLogPath).Last();
+                if (!RefLogEntry.TryParse(refLogContents, out refLog))
+                {
+                    error = $"Last ref log entry for '{fullSymbolicRef}' is unparsable.";
+                    return false;
+                }
+            }
+            catch (IOException ex)
+            {
+                refLog = null;
+                error = $"IOException while reading reflog '{fullSymbolicRef}': " + ex.Message;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryParseRef(string fullSymbolicRef, List<string> messages = null)

```

Comment:
Rename `messages` to `errorMessages` and maybe it should just be and `out string errorMessage` since it doesn't look like we are adding multiple messages?

Reply:
Looks like this could be applied to multiple places in these changes.

Reply:
The `abstract class RepairJob` defines the `HasIssue` and `TryFixIssue` methods and gives you a `List<string>` for all messages to be output as part of the job execution (not just errors).

Keeping the list seems simpler as the caller can just pass in its reference to `messages` to all the called methods. It also avoids the caller(s) having to do a null check for the returned string, then an add to it's list if non-null.

Also I was trying to keep the original implementation of `GitHeadRepairJob` as much as possible since it was pre-existing. 😄